### ISSUE TITLE
Core specification doc needs more useful history

### ISSUE THREAD

**Issue by ghghost** [WebAssembly/design#1440]
The core specification document changes very frequently (often every few weeks), but the History section for version 1.1 only shows the accumulated changes since 1.0. This makes things very difficult for implementors who want to track this specification as it evolves. I advocate strongly that the document include a real incremental history section, with a sub-section for each dated release listing the changes that have been made since the previous dated release.

**Comment by rossberg** [comment]
We do not have anything like "dated releases", I'm afraid, as the spec is supposed to be a "live" document. So such a history would technically be no different from looking at the list of merged PRs or commits.

But most of the changes happening on the repo are minor fixes, editorial tweaks, test additions, or infrastructure adjustments. Normative changes or extensions to the spec occur only by merging proposals, and those are fully covered by the history, as far as they affect the core spec (or should be, please let us know if you find something is missing).

**Comment by ghghost** [comment]
> We do not have anything like "dated releases", I'm afraid, as the spec is
> supposed to be a "live" document. So such a history would technically be no
> different from looking at the list of merged PRs or commits.

In terms of content, I agree.  My concern is with process, not content.

Which of the following is recommended for implementors who want to track the
spec?

* Don't attempt to implement "WebAssembly 1.1" at all until its spec is
frozen.

* Not state that an implementation is of "WebAssembly 1.1," until the spec
is frozen.  (I would expect that in this case code could be developed, but
not usefully distributed outside of the team working on it.)

* State that an implementation is of "WebAssembly 1.1 as of <edition date>,"
assuming that every dated edition (a better term than "release") remains
available.

* Something else?

> But most of the changes happening on the repo are minor fixes, editorial
> tweaks, test additions, or infrastructure adjustments. Normative changes or
> extensions to the spec occur only by merging proposals, and those are fully
> covered by the history, as far as they affect the core spec (or should be,
> please let us know if you find something is missing).

What prompted my comment was the v1.1 editions I've downloaded, dated
2020-07-01, 2021-07-27, 2021-11-03, 2022-01-28, and 2022-02-15.

2020-07-01 has no history section: I didn't go through it in detail to see
whether it actually adds anything to 1.0.

2021-07-27 has a history section, presumably consisting of all the changes
from v1.0 (but no indication as to whether or not these are the same as
the ones in the 2020-07-01 edition, if any).

2021-11-03 has history identical to 2021-07-27.

2022-01-28 adds the vector instructions.

2022-02-15 has history identical to 2022-01-28.

I expect there will be further editions before v1.1 is frozen, some of which
will have only minor tweaks, some of which will include substantial new
material.

My issue as an implementor is that I was hoping that I could look at each
edition of the spec and be able to see immediately what it changed from the
previous one, rather than having to look through a list of PRs or commits,
or compare the history sections by eye.  The current history section
supports this for numbered releases (1.0, 1.1, ...), but not for incremental
editions between releases.  What I was imagining was history looking
something like this (in date order):

	Draft of <date>

	(list of additions / changes)

	Draft of <date>

	(list of additions / changes)

Editions / drafts with no substantive changes don't need history, even if
they change the edition date.

I don't think this adds significant work for someone editing the spec, and I
think it would improve its quality significantly.

Perhaps my expectations, which are based on my recent experience with the
RISC-V ISA spec and to a lesser extent with other evolving technical specs,
are inappropriate?

-- 

L Peter Deutsch ***@***.***> :: Aladdin Enterprises :: Burnaby, BC

**Comment by rossberg** [comment]
Yeah, I understand the dilemma. The CG decided to adopt the "live" document approach. Personally, I think having a proper version history with frequent enough releases would be valuable. But the W3C process for releasing "official" versions is rather slow, so mostly we don't bother.

The document is built automatically with every commit to the repo. There are no explicit releases of editions. The date you see is simply that of the last commit.

My suggestion would be to either say "spec as of <date>", or "version 1.0 + proposals X, Y, and Z", or possibly both.

The Change History section was only added in early 2021. Before that, nobody had thought of it. It should have been up to date since then. It only changes when we merge an actual proposal.
